### ISSUE TITLE
IE: Incorrect printing in Internet Explorer

### ISSUE THREAD

**Issue by dmitry3** [mozilla/pdf.js#3983]
Pdf.js demo is incorrectly printed in IE9-IE11.
OS: Windows 7 64 bit, Windows 8
Here we go:
http://mozilla.github.io/pdf.js/web/viewer.html
Try to print a page in IE 9 with default settings.
Result is horrible:
![pdfjs_portrait](https://f.cloud.github.com/assets/6066311/1644808/87bc3f00-58ee-11e3-9550-8884df0aa8a1.png)

This problem can be solved, if we call "Page Setup" dialog in IE 9 and clear "Enable Shrink-to-Fit" checkbox. By the way, it must be stated in wiki somethere.
Also, when we print a pdf with landscape orientation, pdfjs is not smart enough to rotate canvases in IE (Chrome printing is OK), user must set orientation by hand.

But even with  "Enable Shrink-to-Fit" cleared, printing is incorrect: we have overlapped pages.
Example:
![pdfjs_overlap](https://f.cloud.github.com/assets/6066311/1644849/629b985a-58ef-11e3-8900-b726dcc8aa6e.png)
v0:= ld state[748] is printed twice, and subsequent pages are even more overlapped.

It seems to me, we can fix this problem with style:
<style>
#printContainer div{
    overflow: hidden;
}
</style>

But there is still a problem, we have an empty page printed after all pdf's pages. We could not solve this

**Comment by mattprokes** [comment]
Hi guys, the problem is with 
var PRINT_OUTPUT_SCALE = 2; and for some reason everything is in pt.

For IE change "PRINT_OUTPUT_SCALE" to "1", and change the scaling from "pt" to "px".
It will work.

~matt

**Comment by 111crb111** [comment]
My solution for bad quality, dot know where post it, so post here:

var PRINT_OUTPUT_SCALE = 12;

**Comment by himawan-r** [comment]
I'm using IE 11 and experiencing the same error as @dmitry3 where half to the bottom of the page is missing, i have tried the suggestion above like 

1. change the "pt" to "px" and var cssScale = 'scale(1, 1)'; 
    the result is the printed page smaller than actual size
2. change var PRINT_OUTPUT_SCALE = 12; 
    the result is blank page printed 
3. @rjferguson21  suggestion
    the result is page printed larger than the actual size so that half of the page to the right is missing

Any further information how to solve this ?

**Comment by mattprokes** [comment]
@himawan-r... Please make sure shrink-to-fit is enabled.. it is an IE default, and is likely the reason for the issues.

~matt

**Comment by 111crb111** [comment]
bad print qulity when printing in chrome, and that parametr PRINT_OUTPUT_SCALE  bad solution cause it is need much memory. I know my english very bad :P

**Comment by mattprokes** [comment]
@111crb111  Scaling things 1:1 will likely show the best print quality results. I would recommend giving that a try.

**Comment by himawan-r** [comment]
@mattprokes  i've enabled/disable the shrink to fit, change all "pt" to "px", plus add the scaling thing. Anything i might miss?

**Comment by mattprokes** [comment]
@himawan-r the print scaling is likely the thing that is throwing printing off. Please do not enable/disable shrink-to-fit. Enable it. Since it is printing smaller than the page says to me immediately already that you do not have it enabled, otherwise it would either over-shoot the page or be correct. How are you making these changes? Directly in the source file? Or some other method?

**Comment by himawan-r** [comment]
@mattprokes  i make the change on the viewer.js file on the beforePrint function , yes i mean i tried printing both when i disabled and enabled the shrink-to-fit.

**Comment by himawan-r** [comment]
@mattprokes hi matt, sory for all the problem, your way is indeed did a great job. Silly me there is this margin setting on IE when i set it to 0 now the content is good except for a little blury than what i get from chrome.

**Comment by 111crb111** [comment]
yes chrome, why they is so bad printing quality ? i try all setting in beforePrint func only helps is set high PRINT_OUTPUT_SCALE value but it is need alot memory for printing in chrome, so any ideas ? And one more question why print output looks diferent in chrome and ff when i use same pdfjs!?

**Comment by mattprokes** [comment]
Keep in mind, if print quality if off for a particular browser, there is always browser detection options (PDF.js in fact uses browser detection its' self) . As far as I know the print fidelity is good enough for the company I work at (Merrill Corp), for all browsers. If it is off for chrome; then I recommend only enabling the suggested tweak for IE since as far as I know, this is an IE only problem.

**Comment by himawan-r** [comment]
well i hope there is a better fix for this though, i'm not even sure why page cut off for ie when we set the scaling to 1:2, can someone point me to some article or reading regarding this problem. just curious

**Comment by Jorich** [comment]
Why is this issue closed? Is there a normal fix ?

**Comment by timvandermeij** [comment]
@Jorich This issue is not closed; it is still open because there is no good fix yet.

**Comment by razvar** [comment]
**@mattprokes**  It will be great if you can share the added extensions that allows dynamic watermarking from within the viewer. Thanks a lot.

**Comment by timvandermeij** [comment]
Please don't go off-topic here. This issue is related to printing in IE. Dynamic watermarking is not related.

**Comment by rayzorfin** [comment]
I'm seeing this issue on IE10/11 but not on all the documents. Looking for a fix

**Comment by rayzorfin** [comment]
@mattprokes I didn't have luck with it.
@himawan-r where exactly did you set the shrink-to-fit margins from?

**Comment by TrashbinPornoDolske** [comment]
Usted puede lamer mi Pene y mis Huevos de oro y besar mi Culo hijo de puta con un Trastorno de la personalidad desde el más profundo de los abismos del infierno desde los Abismos de Mozilla!

Vete a la mierda y que tu Madre te autista hijo de puta, vas a lamer Putin Huevos, Puto Maricon Fascista!

La mierda de su Perro, su Gato y su sucio Cabra de montaña que Maricon Pendejo!

**Comment by AssholeDotzler** [comment]
Usted ama el dolor? Y para que te haga un favor... voy a hacer eso!


Yo saque sus huevos de oro de fracaso tan duro hasta que están sonando y que están cantando todas las melodías de dolor hacia adelante y hacia atrás en un instante! Yo soy el cascanueces chico que las grietas de su nueces que nutheads!

Y ahora vete a la mierda ESTÚPIDA de MIERDA COÑO!

----

Usted puede lamer mi Pene y mis Huevos de oro y besar mi Culo hijo de puta con un Trastorno de la personalidad desde el más profundo de los abismos del infierno desde los Abismos de Mozilla!

Vete a la mierda y que tu Madre te autista hijo de puta, vas a lamer Putin Huevos, Puto Maricon Fascista!

La mierda de su Perro, su Gato y su sucio Cabra de montaña que Maricon Pendejo!

Chupar Mozilla Pollas y todos juntos ir y chupar de Google Culo lleno de Mierda!

A la mierda en su rodilla le cock suckers! Y que Carajo Australis Chrome clon de mierda! Follando Mozilla gay lesbiana homosexual hijos de puta del infierno!

**Comment by AssholeDotzler** [comment]
Usted ama el dolor? Y para que te haga un favor... voy a hacer eso!


Yo saque sus huevos de oro de fracaso tan duro hasta que están sonando y que están cantando todas las melodías de dolor hacia adelante y hacia atrás en un instante! Yo soy el cascanueces chico que las grietas de su nueces que nutheads!

Y ahora vete a la mierda ESTÚPIDA de MIERDA COÑO!

----

Usted puede lamer mi Pene y mis Huevos de oro y besar mi Culo hijo de puta con un Trastorno de la personalidad desde el más profundo de los abismos del infierno desde los Abismos de Mozilla!

Vete a la mierda y que tu Madre te autista hijo de puta, vas a lamer Putin Huevos, Puto Maricon Fascista!

La mierda de su Perro, su Gato y su sucio Cabra de montaña que Maricon Pendejo!

Chupar Mozilla Pollas y todos juntos ir y chupar de Google Culo lleno de Mierda!

A la mierda en su rodilla le cock suckers! Y que Carajo Australis Chrome clon de mierda! Follando Mozilla gay lesbiana homosexual hijos de puta del infierno!

**Comment by AssholeDotzler** [comment]
Usted ama el dolor? Y para que te haga un favor... voy a hacer eso!


Yo saque sus huevos de oro de fracaso tan duro hasta que están sonando y que están cantando todas las melodías de dolor hacia adelante y hacia atrás en un instante! Yo soy el cascanueces chico que las grietas de su nueces que nutheads!

Y ahora vete a la mierda ESTÚPIDA de MIERDA COÑO!

----

Usted puede lamer mi Pene y mis Huevos de oro y besar mi Culo hijo de puta con un Trastorno de la personalidad desde el más profundo de los abismos del infierno desde los Abismos de Mozilla!

Vete a la mierda y que tu Madre te autista hijo de puta, vas a lamer Putin Huevos, Puto Maricon Fascista!

La mierda de su Perro, su Gato y su sucio Cabra de montaña que Maricon Pendejo!

Chupar Mozilla Pollas y todos juntos ir y chupar de Google Culo lleno de Mierda!

A la mierda en su rodilla le cock suckers! Y que Carajo Australis Chrome clon de mierda! Follando Mozilla gay lesbiana homosexual hijos de puta del infierno!

**Comment by AssholeDotzler** [comment]
Usted ama el dolor? Y para que te haga un favor... voy a hacer eso!


Yo saque sus huevos de oro de fracaso tan duro hasta que están sonando y que están cantando todas las melodías de dolor hacia adelante y hacia atrás en un instante! Yo soy el cascanueces chico que las grietas de su nueces que nutheads!

Y ahora vete a la mierda ESTÚPIDA de MIERDA COÑO!

----

Usted puede lamer mi Pene y mis Huevos de oro y besar mi Culo hijo de puta con un Trastorno de la personalidad desde el más profundo de los abismos del infierno desde los Abismos de Mozilla!

Vete a la mierda y que tu Madre te autista hijo de puta, vas a lamer Putin Huevos, Puto Maricon Fascista!

La mierda de su Perro, su Gato y su sucio Cabra de montaña que Maricon Pendejo!

Chupar Mozilla Pollas y todos juntos ir y chupar de Google Culo lleno de Mierda!

A la mierda en su rodilla le cock suckers! Y que Carajo Australis Chrome clon de mierda! Follando Mozilla gay lesbiana homosexual hijos de puta del infierno!

**Comment by AssholeDotzler** [comment]
Usted ama el dolor? Y para que te haga un favor... voy a hacer eso!


Yo saque sus huevos de oro de fracaso tan duro hasta que están sonando y que están cantando todas las melodías de dolor hacia adelante y hacia atrás en un instante! Yo soy el cascanueces chico que las grietas de su nueces que nutheads!

Y ahora vete a la mierda ESTÚPIDA de MIERDA COÑO!

----

Usted puede lamer mi Pene y mis Huevos de oro y besar mi Culo hijo de puta con un Trastorno de la personalidad desde el más profundo de los abismos del infierno desde los Abismos de Mozilla!

Vete a la mierda y que tu Madre te autista hijo de puta, vas a lamer Putin Huevos, Puto Maricon Fascista!

La mierda de su Perro, su Gato y su sucio Cabra de montaña que Maricon Pendejo!

Chupar Mozilla Pollas y todos juntos ir y chupar de Google Culo lleno de Mierda!

A la mierda en su rodilla le cock suckers! Y que Carajo Australis Chrome clon de mierda! Follando Mozilla gay lesbiana homosexual hijos de puta del infierno!

**Comment by AssholeDotzler** [comment]
Usted ama el dolor? Y para que te haga un favor... voy a hacer eso!


Yo saque sus huevos de oro de fracaso tan duro hasta que están sonando y que están cantando todas las melodías de dolor hacia adelante y hacia atrás en un instante! Yo soy el cascanueces chico que las grietas de su nueces que nutheads!

Y ahora vete a la mierda ESTÚPIDA de MIERDA COÑO!

----

Usted puede lamer mi Pene y mis Huevos de oro y besar mi Culo hijo de puta con un Trastorno de la personalidad desde el más profundo de los abismos del infierno desde los Abismos de Mozilla!

Vete a la mierda y que tu Madre te autista hijo de puta, vas a lamer Putin Huevos, Puto Maricon Fascista!

La mierda de su Perro, su Gato y su sucio Cabra de montaña que Maricon Pendejo!

Chupar Mozilla Pollas y todos juntos ir y chupar de Google Culo lleno de Mierda!

A la mierda en su rodilla le cock suckers! Y que Carajo Australis Chrome clon de mierda! Follando Mozilla gay lesbiana homosexual hijos de puta del infierno!

**Comment by AssholeDotzler** [comment]
Usted ama el dolor? Y para que te haga un favor... voy a hacer eso!


Yo saque sus huevos de oro de fracaso tan duro hasta que están sonando y que están cantando todas las melodías de dolor hacia adelante y hacia atrás en un instante! Yo soy el cascanueces chico que las grietas de su nueces que nutheads!

Y ahora vete a la mierda ESTÚPIDA de MIERDA COÑO!

----

Usted puede lamer mi Pene y mis Huevos de oro y besar mi Culo hijo de puta con un Trastorno de la personalidad desde el más profundo de los abismos del infierno desde los Abismos de Mozilla!

Vete a la mierda y que tu Madre te autista hijo de puta, vas a lamer Putin Huevos, Puto Maricon Fascista!

La mierda de su Perro, su Gato y su sucio Cabra de montaña que Maricon Pendejo!

Chupar Mozilla Pollas y todos juntos ir y chupar de Google Culo lleno de Mierda!

A la mierda en su rodilla le cock suckers! Y que Carajo Australis Chrome clon de mierda! Follando Mozilla gay lesbiana homosexual hijos de puta del infierno!

**Comment by pudgeball** [comment]
Hey there, I'm wondering if there has been any updates on the issue of IE 11 printing only half the page?

**Comment by mattprokes** [comment]
@pudgeball as far as I know there are no issues with the current solutions provided. If IE is printing 1/2 pages, it is most likely related to "shrink-to-fit" not being enabled. Thus causing the entire page to be offset incorrectly.

**Comment by pudgeball** [comment]
@mattprokes For me (IE version 11.0.9600.17498) printing the demo page (http://mozilla.github.io/pdf.js/web/viewer.html) with Shrink-to-Fit on/off will print half pages.

Is there anything else that I need to change to stop it from occurring?

**Comment by mattprokes** [comment]
@pudgeball yes, the solutions described earlier in the thread will likely solve things for now. The demo page you are talking about does not likely have these changes. As far as something being pushed into mainline which fixes things in a more perm way. Not sure on that one.

**Comment by himawan-r** [comment]
@rayzorfin yes, as @mattprokes  said you need to enable the shrink to fit on IE settings on the browser dude, and this is the margin settings where i set all of them to 0.

![image](https://cloud.githubusercontent.com/assets/8610314/7815560/fe42ca5a-037d-11e5-893d-854f69c79ff3.png)

But still i've got blurry image, still readable though. We still need a better fix for this , and vote for this fix.

**Comment by pudgeball** [comment]
@himawan-r I thought I had set up Shrink to fit and was still seeing the same issue.  I will try this again tomorrow and see if that fixes my problem.

**Comment by pudgeball** [comment]
So I finally was able to test some things on our setup, and this is what I've found.  

Chrome printing of the PDF I was testing worked without any issues.  

IE 11 displayed the PDF correctly however it only printed on half a page at a time.  If I set the PRINT_OUTPUT_SCALE to 1 it would print correctly without being blurry.  

IE 10 displayed the PDF correctly however it also only printed on half of the page.  If I set the PRINT_OUTPUT_SCALE to 1 it would print correctly however it would be blurry.

Was anyone able to find a solution for blurry printing?

**Comment by nicolasboucher** [comment]
After quite some tests, I ended up with a solution that is working fine (at least for me) with IE9-IE11, Firefox and Chrome (didn't test on safari). I'll try to explain the reasoning behind the code, but first, the code (viewer.js : pageViewBeforePrint : ~line 3470) : 

var viewport = pdfPage.getViewport(2); //change
      // Use the same hack we use for high dpi displays for printing to get better
      // output until bug 811002 is fixed in FF.
    var PRINT_OUTPUT_SCALE = 1; //change
    var canvas = document.createElement('canvas');
    canvas.width = Math.floor(viewport.width) * PRINT_OUTPUT_SCALE;
    canvas.height = Math.floor(viewport.height) * PRINT_OUTPUT_SCALE;
    canvas.style.width = (PRINT_OUTPUT_SCALE * Math.floor(viewport.width)) + 'px'; //change
    canvas.style.height = (PRINT_OUTPUT_SCALE * Math.floor(viewport.height)) + 'px'; //change
    var cssScale = 'scale(' + (1 / PRINT_OUTPUT_SCALE) + ', ' +
                              (1 / PRINT_OUTPUT_SCALE) + ')';
    CustomStyle.setProp('transform', canvas, cssScale);
    CustomStyle.setProp('transformOrigin', canvas, '0% 0%');

    var printContainer = document.getElementById('printContainer');
    var canvasWrapper = document.createElement('div');
    canvasWrapper.style.width = Math.floor(viewport.width) + 'px'; //change
    canvasWrapper.style.height = Math.floor(viewport.height) + 'px'; //change
    canvasWrapper.style.pageBreakAfter = 'always'; //add

1- Change pt to px since the canvas size is in px
2- remove fraction with Math.floor (px don't use fractions)
3- Instead of upscaling by 2 with PRINT_OUTPUT_SCALE a render of 1, double the render size from 1 to 2 with pdfPage.getViewport(2)

Hope this can help.

**Comment by timvandermeij** [comment]
If this solution indeed applies for all browsers, could you create a pull request so we can review and test it?

**Comment by StevenDStanton** [comment]
I tested the above solution provided by @nicolasboucher  with a printing issue I am having which only affects IE with extra pages being added.

With the change in code I still have a printing issue in IE 11, Chrome and FF still print fine. While it fixed the pages being cut off mid way through, now I am getting an extra blank page after each printed page

**Comment by FesterCluck** [comment]
In my tests with IE11, after introducing the changes provided by @nicolasboucher into the changeset demonstrated at http://107.21.233.14:8877/397d75075264a93/web/viewer.html all printing issues were resolved. They are also resolved when making those changes in master. I do not see the problem @StevenDStanton described. That being said, the acanvasWrapper.style.pageBreakAfter addition wasn't necessary. In testing with and without, it made no difference in my implementation.

Please test and review these changes among your various issues.

Pull request https://github.com/mozilla/pdf.js/pull/6160

**Comment by himawan-r** [comment]
@FesterCluck the moment you posted the link, i tried to print from the page using IE 11 (not to waste paper i used cute pdf printer > print it as pdf and the result usually the same if we print it directly) but the result is still a cut off page. Can someone re-test it again ? I printscreen the pdf and this is the result for the 1st page 

![image](https://cloud.githubusercontent.com/assets/8610314/8446341/8e3bc4ee-1fdb-11e5-8b00-6dc6abb3580a.png)

**Comment by FesterCluck** [comment]
@himawan-r None of the links I posted would simply allow you to print like that. You'd need to grab the branch at https://github.com/FesterCluck/pdf.js and test from that code until it gets merged.

**Comment by himawan-r** [comment]
@FesterCluck  My bad. i thought "demonstrated" means you already implemented the changes there. I will try to implement the changes and post back here the result. Thanks anyway

**Comment by FesterCluck** [comment]
@himawan-r Yeah, I get that. So, in the interest of making it easy:

http://festercluck.js.org/pdf.js/web/viewer.html

That DOES demonstrate the change.

**Comment by inpercima** [comment]
Tested the link in IE10 and IE11. The document was printed well, no cut offs, no problems.

**Comment by FesterCluck** [comment]
@timvandermeij I've created a pull request. Could you and yours give my revisions a test?

**Comment by uniacid** [comment]
+1 Has this been looked at being merged yet? (after tested successfully of course)

Looks like @FesterCluck changes worked and I might just pull those down for a project of mine where I desperately need IE to work... not the only question is will it work in IE9 :santa:

**Comment by fretman92** [comment]
Printing with Firefox 40 is broken on the original version of PDFJS (text is cropped on the right).
Your [fix](http://festercluck.js.org/pdf.js/web/viewer.html) works better in IE and Chrome, but it's worst with Firefox (about 40% of page cropped)
![image](https://cloud.githubusercontent.com/assets/305637/9714974/4de1504c-555c-11e5-8b1e-c7b4159f5144.png)
Workaround: I choose to activate your fix if userAgent is not Firefox :p

**Comment by mortenbangjustesen** [comment]
This bug has been there for almost two years now. Really no stable solution yet?

**Comment by fretman92** [comment]
Try to preview a printing with latest Chrome (45.0.2454.85 m) on https://mozilla.github.io/pdf.js/web/viewer.html, you will be disappointed more than that :p 

Spoiler: 1 page every 2 is blank... this fix fixes it too! Please merge it :+1:

**Comment by inpercima** [comment]
Please merge this feature. Just do it ;) I think it will be better to merge small thinks step for step...

**Comment by timvandermeij** [comment]
This is about PR #6160 which we cannot merge until we know that it will not regress other browsers. This means that it has to be tested manually on all browsers (Firefox, Chrome, IE and probably Opera/Safari too) and different operating systems (Windows, Linux, OS X) before we can land this. There is no use in quickly merging this and then finding out that it fixes IE, but breaks Firefox or Chrome. If someone is willing to do the test work, please provide your findings on the PR.

**Comment by sent85** [comment]
We tried a fix with respect to the printing issues with IE from PDFJS Viewer, which worked with various types/sizes of document.

Fix specific to IE:
 
canvas.style.width = (PRINT_OUTPUT_SCALE/2 * 100) + '%';
canvas.style.height = (PRINT_OUTPUT_SCALE/2 * 100) + '%';
var cssScale = 'scale(' + (PRINT_OUTPUT_SCALE/2) + ', ' +(PRINT_OUTPUT_SCALE/2) + ')';
canvasWrapper.width = canvas.width;
canvasWrapper.height = canvas.height;

Details on the fix:

canvas.style.width -> reduced it from 200% to 100%.
canvas.style.height -> reduced it from 200% to 100%. 
var cssScale -> reduced it from (2, 2) to (1, 1)
canvasWrapper.width -> set the same as canvas width;
canvasWrapper.height ->set the same as canvas height;

This fix will wrap the 'Canvas' with 'CanvasWrapper' which will be wrapped with 'printContainer' and all were intact and hence all the page content will be printed properly.

Also with this, on testing the quality of the printed content form IE is better than the print from Chrome.

Also to remove the IE browser setting for print margin, set the below css value.

@media print{
body{
margin: 0;
}
}


And to remove header and footer, its left with user.
Browser -> File -> Page Setup (UnCheck Headers and Footers)

Fix tested in the below version:(not validated in any other versions.)
Internet Explorer: 11
PDFJS.Version: 1.1.215

Try this and see if this helps, also requesting for any comments on this for better approach if any.

**Comment by comandante3d** [comment]
Thank you @sent85 your solution worked for us.

We tested in :
     - IE 11.0.9600.16428
     - Chrome 46.0.2
     - Firefox 40.0.3
     - Safari 6.1.6

We only got a small trouble in Safari, it is printing one more blank page.

**Comment by FesterCluck** [comment]
It is my guess that these extra blank page hiccups between browsers may
stem from scrollbar differences and the inherent pixel-based nature of
canvas.

Either that or we've got a tail comma on some array and Safari still cares.

**Comment by saranraju** [comment]
We tried a fix with respect to incorrect printing in IE,Print low quality and also for blank page between each page.
The fix is as follows,

       var pdfPage = this.pdfPage;
      //var viewport = pdfPage.getViewport(1);
	  var viewport = pdfPage.getViewport(2);
      // Use the same hack we use for high dpi displays for printing to get
      // better output until bug 811002 is fixed in FF.
     // var PRINT_OUTPUT_SCALE = 2;
	  var PRINT_OUTPUT_SCALE = 1;
      var canvas = document.createElement('canvas');

      // The logical size of the canvas.
      canvas.width = Math.floor(viewport.width) * PRINT_OUTPUT_SCALE;
      canvas.height = Math.floor(viewport.height) * PRINT_OUTPUT_SCALE;

      // The rendered size of the canvas, relative to the size of canvasWrapper.
      canvas.style.width = (PRINT_OUTPUT_SCALE * 100) + '%';
      canvas.style.height = (PRINT_OUTPUT_SCALE * 100) + '%';

      var cssScale = 'scale(' + (1 / PRINT_OUTPUT_SCALE) + ', ' +
                                (1 / PRINT_OUTPUT_SCALE) + ')';
      CustomStyle.setProp('transform' , canvas, cssScale);
      CustomStyle.setProp('transformOrigin' , canvas, '0% 0%');

      var printContainer = document.getElementById('printContainer');
      var canvasWrapper = document.createElement('div');
    //canvasWrapper.style.width = viewport.width + 'pt';
      canvasWrapper.style.width = '595px';
    //canvasWrapper.style.height = viewport.height + 'pt';
      canvasWrapper.style.height = '842px';
      canvasWrapper.appendChild(canvas);
      printContainer.appendChild(canvasWrapper);


Fix Details;
1.Changed the Viewport value from 1 to 2, which improves the print quality.
2. Instead of setting the viewport width and height to Canvas wrapper, we have set the A4 page size dimension (595 * 842 px).

We have tested the fix in IE9+,Chrome and Firefox. In Safari still a blank page is appended to the end of the print document.

Hope it can fixed if we reduce the viewport height further.

**Comment by StepanMelnik** [comment]
We use the following fix to print pdf reports in IE10 and IE11:

      var viewport = pdfPage.getViewport(1);
      // Use the same hack we use for high dpi displays for printing to get
      // better output until bug 811002 is fixed in FF.
      var PRINT_OUTPUT_SCALE = 2;
      var canvas = document.createElement('canvas');

      // The logical size of the canvas.
      canvas.width = Math.floor(viewport.width) * PRINT_OUTPUT_SCALE;
      canvas.height = Math.floor(viewport.height) * PRINT_OUTPUT_SCALE;

      // The rendered size of the canvas, relative to the size of canvasWrapper.
      if (isIE)
      {
          canvas.style.width = viewport.width + 'px'; //changed
          canvas.style.height = viewport.height + 'px'; //changed
      } else {
          canvas.style.width = (PRINT_OUTPUT_SCALE * 100) + '%';
          canvas.style.height = (PRINT_OUTPUT_SCALE * 100) + '%';
      }

      if (isIE)
      {
    	  var cssScale = 'scale(1, 1)';	// changed
      } else {
    	  var cssScale = 'scale(' + (1 / PRINT_OUTPUT_SCALE) + ', ' +
									(1 / PRINT_OUTPUT_SCALE) + ')';
      }

      CustomStyle.setProp('transform' , canvas, cssScale);
      CustomStyle.setProp('transformOrigin' , canvas, '0% 0%');

      var printContainer = document.getElementById('printContainer');
      var canvasWrapper = document.createElement('div');

      if (isIE)
      {
          canvasWrapper.style.width = viewport.width + 'px';	// changed 'pt' -> 'px'
          canvasWrapper.style.height = viewport.height + 'px';	// changed 'pt' -> 'px'
      } else {
          canvasWrapper.style.width = viewport.width + 'pt';
          canvasWrapper.style.height = viewport.height + 'pt';
      }

**Comment by adilm78** [comment]
Hi, I am using PDFjs 1.3.91 and I am getting the issue of content clipping in IE. does 1.3.91 has any of the fixes posted in this blog? or the fixes were uploaded?

**Comment by pixelbits-mk** [comment]
In order to fix printing issue where only half the page is being printed:

      var PRINT_OUTPUT_SCALE = 4; //change (increase to improve quality)
      var canvas = document.createElement('canvas');

      // The logical size of the canvas.
      canvas.width = Math.floor(viewport.width) * PRINT_OUTPUT_SCALE;
      canvas.height = Math.floor(viewport.height) * PRINT_OUTPUT_SCALE;
      
      // The rendered size of the canvas, relative to the size of canvasWrapper.
      canvas.style.width = '100%'; //change
      canvas.style.height = '100%'; //change

      var cssScale = 'scale(1,1)'; //change

Explanation of changes:

    - Increase the PRINT_OUTPUT_SCALE to improve quality.
    - Change canvas.style.width and canvas.style.height to be 100%.
    - Change cssScale to be (1,1) to preserve scale

Tested and confirmed fix on IE 11.494, Edge 25.1, FF 47.0.1, Chrome 51.0

**Comment by yurydelendik** [comment]
I believe this issue was addressed by #7677. Thanks.

**Comment by jp26jp** [comment]
It's back to only printing half of the page. Here are the version/build numbers:

pdfjsVersion: 1.6.210
pdfjsBuild: 4ce2356

[And a link for proof](https://pdfembedderpremium.worldsecuresystems.com/)

**Comment by timvandermeij** [comment]
The current pre-release (see https://github.com/mozilla/pdf.js/releases) contains more printing fixes. The version you're using doesn't, so please try again with the newest version of PDF.js. If problems still remain, please open a new issue with all required details.

**Comment by pushparaj100** [comment]
I Have tried this url in edge browser https://pdfembedderpremium.worldsecuresystems.com/
The print option is not working good. Its only print entire page not our pdf content. Is there any solution for this?

![pdf_print_edge_issues](https://cloud.githubusercontent.com/assets/8383864/24458643/1c6ebad8-14b7-11e7-880f-b07a6d5b1ecb.jpg)
